master/internal/command/command.go

```diff
@@ -475,17 +586,56 @@ func (c *command) toShell(ctx *actor.Context) *shellv1.Shell {
 	}
 }
 
+func recastMetadataToInt32(metadataValue interface{}) []int32 {
+	// When metadata is JSON-parsed from database map[]interface{}, the types need to be converted
+	// from float64(-ish) to int32 for protobufs.
+	switch metadataValue := metadataValue.(type) {
+	case []int32:
+		return metadataValue
+	case []interface{}:
+		if len(metadataValue) == 0 {
+			return nil
+		}
+
+		result := make([]int32, 0, len(metadataValue))
+
+		switch metadataValue[0].(type) {
+		case int32:
+			for _, el := range metadataValue {
+				result = append(result, el.(int32))
+			}
+		case int:
+			for _, el := range metadataValue {
+				result = append(result, int32(el.(int)))
+			}
+		case float64:
+			for _, el := range metadataValue {
+				result = append(result, int32(el.(float64)))
+			}
+		case nil:
+		default:
+			panic(fmt.Sprintf("expected metadata[0] type: %s", reflect.TypeOf(metadataValue[0])))
+		}
+
+		return result
+	case nil:
+		return nil
+	default:
+		panic(fmt.Sprintf("unexpected metadata type: %s", reflect.TypeOf(metadataValue)))
+	}
+}
+
 func (c *command) toTensorboard(ctx *actor.Context) *tensorboardv1.Tensorboard {
 	state := c.refreshAllocationState(ctx)
 	return &tensorboardv1.Tensorboard{
-		Id:             ctx.Self().Address().Local(),
+		Id:             c.stringID(),
 		State:          state.State.Proto(),
 		Description:    c.Config.Description,
 		StartTime:      protoutils.ToTimestamp(ctx.Self().RegisteredTime()),
 		Container:      state.FirstContainer().ToProto(),
-		ServiceAddress: *c.serviceAddress,
-		ExperimentIds:  c.Metadata["experiment_ids"].([]int32),
-		TrialIds:       c.Metadata["trial_ids"].([]int32),
+		ServiceAddress: c.serviceAddress(),
+		ExperimentIds:  recastMetadataToInt32(c.Metadata["experiment_ids"]),
```

Comment:
what a leaky abstraction. maybe metadata could be a union instead? not blocking at all, just food for thought.

Reply:
@stoksc pushed the static typing. looks much better.